### DISCUSSION TITLE
Jetson Xavier NX on Ubuntu 20.04

### DISCUSSION THREAD

**Discussion by marcelino-pensa** [iRobotEducation/create3_docs#60] [Show and tell]
I am creating this unofficial discussion for those who want to use a `Jetson Xavier NX` with `Ubuntu 20.04` along with the `iRobot Create 3`. I will start with my personal documentation on how I achieved that. If you'd like to follow these steps, do it at your own discretion, as I cannot guarantee that it will work seamlessly. This is my first experience with a Jetson Xavier NX and ROS2, so take my words with a grain of salt.

Any ideas, suggestions and corrections are welcome in this discussion page.

**Who should do this**
- This is mainly useful for those who would like to use python 3, or who wants to run native applications from ROS2 Galactic.
- If you'd prefer to develop natively on Ubuntu, instead of interfacing with the robot through docker, as suggested in [iRobot's documentation](https://iroboteducation.github.io/create3_docs/setup/jetson/). I think it might still be possible to install ROS Foxy on Ubuntu 18.04 and communicate with the robot without docker, but I haven't tested that myself
- Other reasons that other users might point out

**Who should not do this**
Nvidia doesn't have an image with Ubuntu 20.04 for the Jetson boards (if you found one, please add to this discussion). As of last month, it seems like [ubuntu 20.04 is now supported on some Jetson platforms](https://jetsonhacks.com/2022/04/07/jetpack-5-0-developer-preview-release/), but I could not find an image myself. 

**How to install Ubuntu 20.04 on a Jetson Xavier NX**
- Since I couldn't find a Ubuntu 20.04 image for my Jetson Xavier NX, first I followed the instructions from [Nvidia's support](https://developer.nvidia.com/embedded/learn/get-started-jetson-xavier-nx-devkit) to flash Ubuntu 18.04 on this board.

- With Ubuntu 18.04 booted on your Jetson, the next step is to upgrade it to Ubuntu 20.04. Do this at your own risk, but things have been working fine to me. I followed the instructions from [this link](https://qengineering.eu/install-ubuntu-20.04-on-jetson-nano.html). It worked for me, despite being written originally for a Jetson Nano.

**Install ros2 Galactic**
- This step is simple, just follow [iRobot's instructions](https://iroboteducation.github.io/create3_docs/setup/ubuntu2004/). In the RMW step, choose `export RMW_IMPLEMENTATION=rmw_fastrtps_cpp`

**Install Intel Realsense (optional)**
- Run the `buildLibrealsense.sh` script from [this repository](https://github.com/jetsonhacks/installRealSenseSDK). Make sure to use gcc 8 for that, and switch to gcc 9 afterwards (see [this link](https://qengineering.eu/install-ubuntu-20.04-on-jetson-nano.html) again for reference).

**Test that you can communicate with the Create 3 robot**
- Make sure that the robot is connected to the same wifi network as you are. See [this link](https://edu.irobot.com/create3-setup) for reference
- Run the command `ros2 topic list`, you should see the following:
```
/battery_state
/cmd_audio
/cmd_lightring
/cmd_vel
/dock
/hazard_detection
/imu
/interface_buttons
/ir_intensity
/ir_opcode
/kidnap_status
/mouse
/odom
/parameter_events
/robot_state/transition_event
/rosout
/slip_status
/static_transform/transition_event
/stop_status
/system_monitor/transition_event
/tf
/tf_static
/wheel_status
/wheel_ticks
/wheel_vels
```

Finally, test that you can get data from the robot by running `ros2 topic echo /battery_state`. You should see something like this:
```
header:
  stamp:
    sec: 1652920296
    nanosec: 700261009
  frame_id: ''
voltage: 14.432000160217285
temperature: 30.149999618530273
current: -0.2280000001192093
charge: 0.8299999833106995
capacity: 2.049999952316284
design_capacity: 2.049999952316284
percentage: 0.4000000059604645
power_supply_status: 0
power_supply_health: 0
power_supply_technology: 0
present: true
cell_voltage: []
cell_temperature: []
location: ''
serial_number: ''
---
```

**Using your keyboard to teleoperate the robot around**
- Open a terminal in the Jetson and install the keyboard teleop node: `sudo apt install ros-galactic-teleop-twist-keyboard`
- Run the node: `ros2 run teleop_twist_keyboard teleop_twist_keyboard`
- Now you'll be able to drive the robot around using the keyboard keys described in the terminal

**Dock/Undock**
- If the robot is docked, you can request it to undock with the command: `ros2 action send_goal /undock irobot_create_msgs/action/Undock "{}"`
- If the robot is undocked, you can request it to dock with the command: `ros2 action send_goal /dock irobot_create_msgs/action/DockServo "{}"`

**Comment by marcelino-pensa** [comment]
**Bridge between ROS1 and ROS2**
If you already have code written in ROS1 and want to test it with the robot, you will need to use the `ros1_bridge`.

- First, create a ROS2 workspace [as described in here](https://docs.ros.org/en/galactic/Tutorials/Workspace/Creating-A-Workspace.html)
- Navigate to the `src` folder of your ROS2 workspace, and clone the following repositories:
  - `git clone git@github.com:ros2/ros1_bridge.git`
  - `git clone git@github.com:iRobotEducation/irobot_create_msgs.git`
- Follow the installation instructions in the Readme file of [ros1_bridge](https://github.com/ros2/ros1_bridge). Grab a coffee, as this step took over 30min to me.
- Follow the instructions in the Readme file to test `ros1_bridge`

**Test robot keyboard teleoperation using ROS1`**
You'll need 3 terminals:
- In the first terminal, start roscore: `source ${ROS1_INSTALL_PATH}/setup.bash && roscore`
- Start the bridge in the second terminal: `source ${ROS1_INSTALL_PATH}/setup.bash && source ${ROS2_INSTALL_PATH}/setup.bash && ros2 run ros1_bridge dynamic_bridge`
- Start ROS1 teleop node on a third terminal: `source ${ROS1_INSTALL_PATH}/setup.bash && rosrun teleop_twist_keyboard teleop_twist_keyboard.py`

**Comment by bluecamel** [comment]
Thanks for sharing!  I almost went this direction before discovering the [docker containers](https://github.com/dusty-nv/jetson-containers), but I'm going to have to try this now.

**Comment by eecue** [comment]
This was super helpful thanks!

**Comment by eecue** [comment]
Actually, I noticed the when running `realsense-viewer` it's not using any GPU ... is yours working? did you need to do anything special to get it to work?

**Reply by marcelino-pensa** [reply]
I actually never paid attention to whether `realsense-viewer` uses any GPU, but I recall using `realsense-viewer` with no issue. Note that I only used it with a T265, I did not spend any time testing other realsense devices

**Comment by ParadoxRobotics** [comment]
How have you install the following package : https://github.com/iRobotEducation/irobot_create_msgs ?
via Debian or from source ?
